llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp

```diff
@@ -1971,20 +1976,37 @@ maybeMoveCFI(MachineInstr &MI, MachineBasicBlock::iterator MaybeCFI) {
   }
 }
 
-MachineBasicBlock::iterator
-AArch64LoadStoreOpt::mergeUpdateInsn(MachineBasicBlock::iterator I,
-                                     MachineBasicBlock::iterator Update,
-                                     bool IsPreIdx) {
+std::optional<MachineBasicBlock::iterator> AArch64LoadStoreOpt::mergeUpdateInsn(
+    MachineBasicBlock::iterator I, MachineBasicBlock::iterator Update,
+    bool IsForward, bool IsPreIdx, bool MergeEither) {
   assert((Update->getOpcode() == AArch64::ADDXri ||
           Update->getOpcode() == AArch64::SUBXri) &&
          "Unexpected base register update instruction to merge!");
   MachineBasicBlock::iterator E = I->getParent()->end();
   MachineBasicBlock::iterator NextI = next_nodbg(I, E);
 
-  // If updating the SP and the following instruction is CFA offset related CFI
-  // instruction move it after the merged instruction.
-  MachineBasicBlock::iterator CFI =
-      IsPreIdx ? maybeMoveCFI(*Update, next_nodbg(Update, E)) : E;
+  // If updating the SP and the following instruction is CFA offset related CFI,
+  // make sure the CFI follows the SP update either by merging at the location
+  // of the update or by moving the CFI after the merged instruction. If unable
+  // to do so, bail.
+  MachineBasicBlock::iterator InsertPt = I;
+  if (IsForward) {
+    assert(IsPreIdx);
+    if (auto CFI = maybeMoveCFI(*Update, next_nodbg(Update, E)); CFI != E) {
```

Comment:
I don't think I've seen this pattern here. Does the statement only pass if `CFI` is nonnull and `CFI != E`?

For clarity, can we split this into a separate statement? Then I guess the if statement would be simply `CFI && CFI != E`.

Reply:
Do you mean an `if` statement with an initializer? The `C++17` standard was only adopted in LLVM [two years ago](https://reviews.llvm.org/D130689), so this usage may not be very common yet, but it is still quite normal.

Reply:
I've seen the initializer in an if statement, but I don't think I've seen a `;` inside an if statement

Reply:
Still, there are already quite a few such statements throughout the LLVM codebase.

Reply:
Interesting! I didn't know that was a feature.
https://skebanga.github.io/if-with-initializer/

Although I wish there was a comment on this in the coding standards.
https://llvm.org/docs/CodingStandards.html

Reply:
It seems to be used in just a few places.
https://github.com/search?q=%2Fif+%5C%28.*%3B.*%5C%29%2F+repo%3Allvm%2Fllvm-project+path%3Allvm%2Flib&type=code